### ISSUE TITLE
[html] New HTML language support should allow files with extensions other than `html`

### ISSUE THREAD

**Issue by SCWells72** [pmd/pmd#3952]
I'm integrating the new HTML language support from 6.45.0 into my IDE plugin (Illuminated Cloud 2) and noticed that those rules are only active for files with the `html` file extension as evidenced by how `HtmlLanguageModule` chains to its superclass' constructor with only that file extension. My plugin executes PMD using `PMD#processFiles` with a `ReaderDataSource` that created against the original source file's body and path. When the filename in that path has an extension other than HTML, e.g., `page`, `component`, `cmp`, etc. (which are all inherently HTML markup with an extended tag set), the HTML rules don't execute.

Is there perhaps there's some way for me to register other file extensions for a specific language? I didn't see anything obvious looking at how `LanguageRegistry#findByExtension` works, but hopefully I just missed something.

**Comment by SCWells72** [comment]
Note that I've found `PMDConfiguration#setForceLanguageVersion`, but if I explicitly set that to HTML then other relevant rules for that file type aren't executed, specifically for hybrid files like [Visualforce](https://pmd.github.io/latest/pmd_rules_vf.html) which should be evaluated as both Visualforce and HTML.

**Comment by adangel** [comment]
There are different ways to go forward (I don't guarantee that any of these work, these are just ideas):

* You are integrating html not to analyze vanilla html files, but to analyze salesforce flavors of it. As discussed in #3846 it would make sense to define an own language (dialect?) that reuses the html module. There you could define your own file extensions. That has also the advantage, that the (hopefully soon suggested) rules are not poising the html ruleset - I expect, that the rules you want to implement, are very very salesforce specific.
* We could - as you suggested - create an API, so that you can modify the recognized file extensions for a language module. This would the the programmatic way for #461. But that would mean just that: you are forcing the html module to be used against salesforce files. As soon as the first rule is suggested and it is salesforce specific, we won't add it to the html module, but demand a salesforce specific language module.
However - this might be a easier way to gain experience with this new possibility.
* Depending on how you call PMD, you could use the `FileCollector` API. See variant 2 in https://pmd.github.io/latest/pmd_userdocs_tools_java_api.html. It should be possible to manually add the file with a specific language:

```
var htmlModule = LanguageRegistry.findLanguageByTerseName("html");
...
pmd.files().addFile(Paths.get("pages", "MyPage.page"), htmlModule);
```

Looking at the code, I think, this only works in PMD 7 - in pmd 6 we loose the language info... We might just fix that somehow (maybe storing a explicit mapping of the pathname to language for later). Given this is an API that's forward compatible, it would be the preferred solution.
* You could create a sub class of `PMDConfiguration` and overwrite `getLanguageVersionOfFile()`:

https://github.com/pmd/pmd/blob/eae87d26a7bce1ae09e841130de320f635d48921/pmd-core/src/main/java/net/sourceforge/pmd/PMDConfiguration.java#L326-L334

This method is called while processing the files and if no forced language is provided. Instead of using the language discoverer, you could check yourself the extension and return the appropriate language. However, this is not forward compatible and is more like a hack.

**Comment by adangel** [comment]
Referring just to the title of this issue: The following extensions are missing:

* `xhtml`
* `htm`

**Comment by adangel** [comment]
> specifically for hybrid files like [Visualforce](https://pmd.github.io/latest/pmd_rules_vf.html) which should be evaluated as both Visualforce and HTML.

@SCWells72 

Can you give a specific example of such files? This is IMHO a complete new use case and not supported at all...
We would need a very concrete example - ideally with an example of what you want to detect in those files. Thanks.

**Comment by SCWells72** [comment]
@adangel yes, they're hybrid in the same sense that any HTML-based templating language is hybrid. There's an "outer" language that's HTML (with its ability to host other languages such as JavaScript and CSS), there's an "inner" language used for expressions, in this case `{!expression}`, and there are additional tags (effectively UI components) that can be used in the HTML markup.

Now, having said that, I'm fine if you want to resolve this as I've already implemented a successful working for it, e.g.:

![image](https://user-images.githubusercontent.com/7671043/169536106-3f10bec9-f84a-4fb9-b7b7-d1721da44535.png)

Basically when I see that I'm in one of these HTML-based templating language contexts, I execute PMD twice on the same file body, once with a path that has the file's actual extension, e.g., `.page`, and one with a `.html` extension, each run with only the applicable rules for that language active, then I adapt the results of those two runs into the IDE's native code inspection results.

So I think my specific use case is covered, albeit with the need for a small workaround. If this isn't a problem for anyone else, I think it can be closed.
